### DISCUSSION TITLE
Mapping Builder: 3rd mapping option

### DISCUSSION THREAD

**Discussion by igarashitm** [atlasmap/atlasmap#3172] [Design & Features]
We've been working on enabling custom mapping in AtlasMap, named `MappingBuilder` (#1514). It's a 3rd option of how user can add a mapping.
### Mapping Options
| | UI assist | Flexibility | Status |
|-|-------------|--------------|-----------|
| DnD Mapping | Full drag and drop mapping with a line drawn between mapped fields. Transformation could be added field by field in UI | Limited to simple mapping, oen-to-one, one-to-many, many-to-one, for-each(collection) | Supported |
| [Conditional Mapping (expression)](https://docs.atlasmap.io/#applying-conditions-to-mappings) | The expression is editable in the UI with field completion assist. A line drawn between mapped fields. Transformation could be applied as a function. Function completion is under development (#2403). <br> Examples in Mapping Definition: [transformation](https://github.com/atlasmap/atlasmap/blob/master/lib/itests/core/src/test/resources/mappings/atlasmapping-expression2-action.json#L18), [filter-select](https://github.com/atlasmap/atlasmap/blob/master/lib/itests/core/src/test/resources/mappings/atlasmapping-expression-filter-select-java.json#L18), [repeat-count](https://github.com/atlasmap/atlasmap/blob/master/lib/itests/core/src/test/resources/mappings/atlasmapping-multiplicity-transformation-action_repeat.json#L22) | Conditional logic could be implemented inline with its spreadsheet math formula like syntax. Most complex mapping logic would not be able to be implemented. | Tech Preview |
| Custom Mapping (#1514) | User implements their own `MappingBuilder` in Java, usually by extending [DefaultAtlasMappingBuilder](https://github.com/atlasmap/atlasmap/blob/master/lib/core/src/main/java/io/atlasmap/builder/DefaultAtlasMappingBuilder.java), create a jar from it and import the jar file into AtlasMap UI. Then user can only enable/disable the custom mapping in the UI. No mapping details could be edited in the AtlasMap UI. | Ultimately any mapping logic that could be written in Java could be achieved. | In Development|

The Custom Mapping is a last resort to add a mapping. The easier way with more UI assist is always better where it's possible. On the other hand, we don't want to make AtlasMap UI too complex to support all the usecases in the UI. DnD Mapping has to stay intuitive, easy and quick. We introduced Conditional Mapping to support advanced usecases without compromising simpleness of the DnD Mapping, and Custom Mapping is to support even more complex usecases that is not possible even with Conditional Mapping. Keep simple things simple and complex things possible.

### Current Status of Custom Mapping
We started from this runtime PoC - #2678
The custom mapping would look like this - `DefaultAtlasMappingBuilder` is in `io.atlasmap:atlas-core`:
```java
import io.atlasmap.builder.DefaultAtlasMappingBuilder;
public class MyDefaultAtlasMappingBuilder extends DefaultAtlasMappingBuilder {
    @Override
    public void processMapping() throws Exception {
        read("SourceDoc1", "/f1") // read "/f1" field from "SourceDoc1" source Document, returns AtlasField
            .write("TargetDoc", "/f2"); // write "/f2" field on "TargetDoc" target Document
        readConstant("c3")
            .writeProperty("p4");
        readProperty("SourceDoc1", "p5") // read a property scoped to SourceDoc1
            .write("TargetDoc", "/f6");
    }
};
```
Just read and write as a first step. It's not useful in practice yet. It should also be able to write a value directly without reading any source, i.e. have `write()` method directly on `MappingBuilder` but not only on `AtlasField` (#3177).

It's not yet supported to enable Custom Mapping from UI (#2679). In order to try it, you'd need to hand edit the mapping definition and add a Custom Mapping entry.
```json
...
"mappings" : {
      "mapping" : [ {
        "jsonType" : "io.atlasmap.v2.CustomMapping",
        "className" : "io.atlasmap.itests.core.MappingBuilderTest$JsonXmlBuilder"
      } ]
    },
...
```

### Planned enhancements
- #2683 
`MappingBuilder` provides `read()` to read a source field and return as an `AtlasField`. Expecting it to also work for reading 
COMPLEX parent field, `AtlasField` should provide a way to access its children.
```java
AtlasField contact = read("SourceDoc1", "/contact");
if (contact.get("language").value().equals("English") {
    write("TargetDoc1", "/greeting", String.format(
        "Hello %s %s", contact.get("firstName").value(), contact.get("lastName").value()));
}
```
It would be just a direct child by specifying the name, but not traveling 2 or more depth like `contact.get("address/addressLine1")`. This could be instead `contact.get("address").get("addressLine1")`
- #2681 
Iterating on collection items is a basic feature we need. 2 options in my mind, we might want both.
```java
AtlasCollectionField collection = readCollection("SourceDoc1", "/contactList<>");
```
```java
AtlasCollectionField collection = read("SourceDoc1", "/contactList<>").collection();
```
where `AtlasCollectionField` would look like
```java
public class AtlasCollectionField extends AtlasField {
    AtlasField get(int index); // Return an indexed item, e.g. /contactList<0>
    AtlasCollectionField get(String path); // Return a collection of the child, e.g. /contactList<>/firstName
    int size();
    List<AtlasField> elements();
```
- #2682 
We already have a bunch of transformations OOTB and even custom transformation is allowed. Custom Mapping also has to leverage them.
```java
    AtlasField action(Action actionModel, AtlasField... field);
```
e.g.
```java
AtlasField firstName = read("SourceDoc1", "/firstName");
AtlasField lastName = read("SourceDoc1", "/lastName");
Concatenate action = new Concatenate();
action.setDelimiter(":");
AtlasField processed = action(action, firstName, lastName);
```

### Thoughts?
Please add your opinion here to improve this feature.

**Comment by alagusundarams** [comment]
We need the repeat_count to support a constant field as well, right now Repeat(count(sourceArrayField), sourceSimpleField) sourceSimpleFields should also come from constants which we define. this will help some of the constant mapping to a specific array filed

**Comment by alagusundarams** [comment]
@shanthi4647, @rkorytkowski  tagging you both for more inputs because i saw the PR https://github.com/atlasmap/atlasmap/pull/2049 raised by @rkorytkowski  was not merged into as the UI got moved from angular to react, but we need similar kind of mapping to be achieved based on a lenght of an array we would like to equate/map a constant value to a specific field in a Array
